x/gamm/twap/store_test.go

```diff
@@ -0,0 +1,181 @@
+package twap_test
+
+import (
+	"time"
+
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap/types"
+)
+
+// TestTrackChangedPool takes a list of poolIds as test cases, and runs one list per block.
+// Every simulated block, checks that there no changed pools.
+// Then runs k.trackChangedPool on every item in the test case list.
+// Then checks that changed pools returns the list, deduplicated.
+//
+// This achieves testing the functionality that we depend on, that this clears every end block.
+func (s *TestSuite) TestTrackChangedPool() {
+	tests := map[string][]uint64{
+		"single":         {1},
+		"duplicated":     {1, 1},
+		"four":           {1, 2, 3, 4},
+		"many with dups": {1, 2, 3, 4, 3, 2, 1},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			// Test that no cumulative ID registers as tracked
+			changedPools := s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Empty(changedPools)
+
+			// Track every pool in list
+			cumulativeIds := map[uint64]bool{}
+			for _, v := range test {
+				cumulativeIds[v] = true
+				s.twapkeeper.TrackChangedPool(s.Ctx, v)
+			}
+
+			changedPools = s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Equal(len(cumulativeIds), len(changedPools))
+			for _, v := range changedPools {
+				s.Require().True(cumulativeIds[v])
+			}
+			s.Commit()
+		})
+	}
+}
+
+// TestGetAllMostRecentRecordsForPool takes a list of records as test cases,
+// and runs storeNewRecord for everything in sequence.
+// Then it runs GetAllMostRecentRecordsForPool, and sees if its equal to expected
+func (s *TestSuite) TestGetAllMostRecentRecordsForPool() {
+	tPlusOne := baseTime.Add(time.Second)
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tPlusOneRecord := newEmptyPriceRecord(1, tPlusOne, "tokenB", "tokenA")
+	tests := map[string]struct {
+		recordsToSet    []types.TwapRecord
+		poolId          uint64
+		expectedRecords []types.TwapRecord
+	}{
+		"set single record": {
+			[]types.TwapRecord{baseRecord},
+			1,
+			[]types.TwapRecord{baseRecord},
+		},
+		"query non-existent pool": {
+			[]types.TwapRecord{baseRecord},
+			2,
+			[]types.TwapRecord{},
+		},
+		"set two records": {
+			[]types.TwapRecord{baseRecord, tPlusOneRecord},
+			1,
+			[]types.TwapRecord{tPlusOneRecord},
+		},
+		"set two records, reverse order": {
+			// The last record, independent of time, takes precedence for most recent.
+			[]types.TwapRecord{tPlusOneRecord, baseRecord},
+			1,
```

Comment:
consider making a record with a poolId other than 1 and having 1 test case for that

The motivation is to have a sanity check that we don't have a dependency on any specific pool id (1 in this case)

Reply:
Similar suggestion for `TestGetRecordAtOrBeforeTime`